extensions/amp-pan-zoom/1.0/component-ts.tsx

```diff
@@ -0,0 +1,268 @@
+import {forwardRef} from '#preact/compat';
+import {ContainWrapper} from '#preact/component';
+import {useResizeObserver} from '#preact/hooks/useResizeObserver';
+
+import {useStyles} from './component.jss';
+import {usePanZoomState} from './hooks/usePanZoomState';
+
+const DOUBLE_TAP_TIME = 500; // Maximum time between tap starts
+
+const gestureConfig: Omit<UserGestureConfig, 'target'> = {
+  drag: {
+    filterTaps: true,
+    tapsThreshold: 10,
+  },
+};
+
+export type BentoPanZoomProps = {
+  children?: ComponentChildren;
+  controls?: boolean;
+  initialScale?: number | string;
+  initialX?: number | string;
+  initialY?: number | string;
+  maxScale?: number | string;
+};
+
+export type BentoPanZoomApi = {
+  transform(scale: number, x: number, y: number): void;
+};
+
+function getElementPosition(
+  clientX: number,
+  clientY: number,
+  element: HTMLElement
+) {
+  const elBounds = element./* REVIEW */ getBoundingClientRect();
+  return {
+    anchorX: clientX - elBounds.x,
+    anchorY: clientY - elBounds.y,
+  };
+}
+
+function classNames(...args: Array<string | false | null | 0>) {
+  return args.filter(Boolean).join(' ');
+}
+
+/**
+ * @return {PreactDef.Renderable}
+ */
+export function BentoPanZoomWithRef(
+  props: BentoPanZoomProps,
+  ref: Ref<BentoPanZoomApi>
+) {
+  const {
+    children,
+    controls = true,
+    initialScale,
+    initialX,
+    initialY,
+    maxScale,
+    ...rest
+  } = props;
+  const styles = useStyles();
+
+  const [state, actions] = usePanZoomState({
+    initialX,
+    initialY,
+    initialScale,
+    maxScale,
+  });
+  const isPannable = state.scale !== 1;
+  const canZoom = state.scale < state.maxScale;
+
+  const contentRef = useRef<HTMLDivElement>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const updateSizes = () => {
+    const contentBox = contentRef.current!./* REVIEW */ getBoundingClientRect();
+    const containerBox =
+      containerRef.current!./* REVIEW */ getBoundingClientRect();
+
+    actions.updateBounds({
+      containerSize: containerBox,
+      contentSize: contentBox,
+      contentOffset: {
+        x: contentBox.x - containerBox.x,
+        y: contentBox.y - containerBox.y,
+      },
+    });
+  };
+  useResizeObserver(contentRef, updateSizes);
+  useResizeObserver(containerRef, updateSizes);
+
+  useImperativeHandle(
+    ref,
+    () => /** @type {!BentoPanZoom.PanZoomApi} */ ({
+      transform: (scale: number, x: number, y: number) => {
+        actions.transform({
+          scale,
+          posX: x,
+          posY: y,
+        });
+      },
+    }),
+    [actions]
+  );
+
+  const panZoomStyles = {
+    transformOrigin:
+      px(-state.contentOffset.x) + ' ' + px(-state.contentOffset.y),
+    transform: translate(state.posX, state.posY) + cssScale(state.scale),
+  };
+
+  const initialStateRef = useRef(state);
+  const lastTapTime = useRef(-DOUBLE_TAP_TIME);
+  const handlers: Partial<UserHandlers> = {
```

Comment:
handlers seems like a good candidate to memoize in `useMemo`. Seems like it's using a lot.

Reply:
This is a tricky one. 
The `useGesture` hook does not have any performance optimizations for a memoized `handlers` object, so there's not a "big win" at stake here.  So if we were to memo it here, we'd save the creation of an object with 6 functions.  
But, some of these handlers need access to local `state`, which makes memoization harder. We'd have to refactor to use a `useValueRef(state)` to access the state, which isn't that hard, but the code quality would suffer.

I'm on the fence on this one ... object creation, versus code quality ... what do you think?

Reply:
Yeah, let's just leave it as is. I think you're probably right. There are to many external dependencies on variables that'll probably just be being rebuild all the time anyway, defeating the entire point.